### DISCUSSION TITLE
How/when are workspaces made?

### DISCUSSION THREAD

**Discussion by bvankuik** [nikitabobko/AeroSpace#875] [questions-and-answers] [answered]
I've got shortcut keys set up for workspaces 1 to 9, like so:

    ctrl-1 = 'workspace 1'
    ctrl-2 = 'workspace 2'
    ...

Does this make workspaces? Or are they rather created when you switch to them? I'm struggling with the mental model.

**Comment by nikitabobko** [comment, marked as answer]
Yes, the mental model is not described in the docs unfortunately. It's somewhere in my todo to write it down at some point.

The general idea is that you don't need to create workspaces explicitly, which should reduce the hussle. You just say `workspace XXX` and AeroSpace will auto-create `XXX` and jump to it.

The workspaces are automatically destroyed when they are not visible, empty, and they are not mentioned in the config.

The last part "are not mentioned in the config" is a bit wonky and wasn't well thought of. If I was doing it over again, I'd reconsider. If I remember it correctly `alt-1 = 'exec-and-forget aerospace workspace 1'` - this doesn't count as "mentioned in the config", but this does: `alt-1 = 'workspace 1'`. And given possibility of #278 and friends, the definition of "mentioned in the config" becomes even more wonkier.

I was also thinking to introduce the following config option at some point:
```toml
# The current "wonky" default
force-keep-workspaces-alive = 'auto-infer-from-config'
# or
force-keep-workspaces-alive = ['1', '2', '3', '4']
```

The only significant difference known to me between alive and dead workspaces is that alive workspaces preserve their attachment to monitors, while dead don't.

In comparison, in i3, workspaces are alive only when they are either non-empty, or visible. And I didn't like that I couldn't reliably move an empty workspace to another monitor and be sure that it would stay there

**Reply by mpeterson** [reply]
Another difference is that an alive workspace will also keep the layout, no? I'd personally would not want to destroy a workspace, even if empty and dynamically created just because of this.

**Reply by nikitabobko** [reply]
> Another difference is that an alive workspace will also keep the layout, no?

Good point, thanks, you're correct. The layout of the root tiling node is preserved for alive workspaces indeed.

> I'd personally would not want to destroy a workspace, even if empty and dynamically created just because of this.

I agree that this use case isn't well considered either

**Reply by nikitabobko** [reply]
https://github.com/nikitabobko/AeroSpace/issues/999
